From a9a17c3221331305199c7390ff528b7d60ef6c08 Mon Sep 17 00:00:00 2001
From: Robert S Peterson <rpeterso@redhat.com>
Date: Wed, 21 May 2014 17:29:10 -0400
Subject: [fs] gfs2: Increase the max number of ACLs

Message-id: <451511823.9721934.1400693350822.JavaMail.zimbra@redhat.com>
Patchwork-id: 80073
O-Subject: [RHEL7.1 PATCH][bz1078231] GFS2: Increase the max number of ACLs
Bugzilla: 1078231
RH-Acked-by: Abhijith Das <adas@redhat.com>
RH-Acked-by: Steven Whitehouse <swhiteho@redhat.com>

Hi,

This is a RHEL7.1 back-port of a patch that's already gone into other
releases. The upstream patch is here:

http://git.kernel.org/cgit/linux/kernel/git/steve/gfs2-3.0-nmw.git/commit/fs/gfs2?id=b00263d1cafdd667de56cde47f35d5ee8dd37e14

It was tested on system gfs-i24c-01.mpc.lab.eng.bos.redhat.com with
a reliable reproducer.

This patch increases the maximum number of ACLs from 25 to 300 for
a 4K block size. The value is adjusted accordingly if the block size
is smaller. Note that this is an arbitrary limit with a performance
tradeoff, and that the physical limit is slightly over 500.

rhbz#1078231

Regards,

Bob Peterson
Red Hat File Systems

Signed-off-by: Bob Peterson <rpeterso@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/gfs2/acl.c b/fs/gfs2/acl.c
index f69ac0a..533d56f 100644
--- a/fs/gfs2/acl.c
+++ b/fs/gfs2/acl.c
@@ -263,7 +263,7 @@ static int gfs2_xattr_system_set(struct dentry *dentry, const char *name,
   goto out_release;
 
  error = -EINVAL;
- if (acl->a_count > GFS2_ACL_MAX_ENTRIES)
+ if (acl->a_count > GFS2_ACL_MAX_ENTRIES(GFS2_SB(inode)))
   goto out_release;
 
  if (type == ACL_TYPE_ACCESS) {
diff --git a/fs/gfs2/acl.h b/fs/gfs2/acl.h
index 0da38dc..ddc2cc8 100644
--- a/fs/gfs2/acl.h
+++ b/fs/gfs2/acl.h
@@ -14,7 +14,7 @@
 
 #define GFS2_POSIX_ACL_ACCESS  "posix_acl_access"
 #define GFS2_POSIX_ACL_DEFAULT  "posix_acl_default"
-#define GFS2_ACL_MAX_ENTRIES  25
+#define GFS2_ACL_MAX_ENTRIES(sdp) ((300 << (sdp)->sd_sb.sb_bsize_shift) >> 12)
 
 extern struct posix_acl *gfs2_get_acl(struct inode *inode, int type);
 extern int gfs2_acl_create(struct gfs2_inode *dip, struct inode *inode);
-- 
1.7.1